### PR TITLE
Cleaner: `keep_files` should only apply to the beginning of paths, not substrings with index > 0

### PR DESCRIPTION
**Pull request by shinkondo**
`keep_files` could prevent more files than intended from being removed. 

For example, if someone is working under a folder something like `username.github.io`, `keep_files:   [".git", ".svn"]` matches any files under `username.github.io/_sites`, leaving the destination folder uncleaned.

This would be confusing when a file is renamed while it is viewed on a browser. An user can be unaware of the fact that he is looking at outdated file (see #3847).

According to the [documentation](http://jekyllrb.com/docs/configuration/#global-configuration), the paths in `keep_files` should be "relative to the destination" rather than being keywords.

For these reasons, I made a change so that keep_files are interpreted as paths under the destination folder.

### CODE REVIEW

**Comment by envygeeks** [comment]
What is the status of this?

**Comment by parkr** [comment]
> For example, if someone is working under a folder something like username.github.io, keep_files: [".git", ".svn"] matches any files under username.github.io/_sites, leaving the destination folder uncleaned.

The example here doesn't apply to the feature. `keep_files` is a way of inhibiting the removal of files and folders inside of your compiled site directory (e.g. `_site`). In your example, it sounds like you think of `keep_files` as a way to keep your site's source from being cleaned up.

Am I misreading you here?

**Comment by shinkondo** [comment]
First, thank you for your feedback.

I am afraid you are. I think of `keep_files` as a way to keep a compiled site directory, not source.
If you would take a look at #3847, you might get some ideas about my intent; I was confused by the fact that compiled outdated html was not removed.  

Let me know if you are still unsure about my post. I know I am not a good writer.

File: test/test_cleaner.rb [hunk 1]
```diff
@@ -37,6 +37,40 @@ class TestCleaner < JekyllUnitTest
...
+      @site = fixture_site
+      @site.keep_files = ['.git']
+
+      @cleaner = Cleaner.new(@site)
+      @cleaner.cleanup!
+    end
+
+    teardown do
+      FileUtils.rm_rf(dest_dir('.git'))
+      FileUtils.rm_rf(dest_dir('.username.github.io'))
```
Comment (parkr) [inline test/test_cleaner.rb]: This should not include the preceding `.`, correct?

Comment (shinkondo) [inline test/test_cleaner.rb]: That is correct. Thank you for pointing it out.

**Comment by parkr** [comment]
@shinkondo I see, thank you for your patience.

> my blog folder name containing ".git" as a substring

This seems to have explained it for me. You have `keep_files: ".git"` with a subdirectory of `username.github.io` and you find that this subdirectory, once created, is not cleaned?

I am 👍 on this PR. Just one comment above.

**Comment by shinkondo** [comment]
I think what you described captures my issue, but please don't pull this request as of now.

It seems to me that the situation has changed because of bd2c337e5bd1a3532f42776f55f80f4e273cfeac. My code in pull-request would not work. Now regex generated from `keep_files` would match against relative paths under `_sites`, not absolute paths. My code will generate regex with absolute paths. I bet it would not match anything practically.

**Comment by parkr** [comment]
@shinkondo You're welcome to merge `origin/master` into this branch and let Travis CI take a crack at it. bd2c337 doesn't fix your issue, though, does it?

**Comment by shinkondo** [comment]
No, bd2c337e5bd1a3532f42776f55f80f4e273cfeac does not fix my issue.

As Travis CI showed, I was wrong to say that my code would not work. It seems that the code is using absolute paths again because of bd2c337e5bd1a3532f42776f55f80f4e273cfeac.

In any case, I can fix my mistake in the test. Is there anything else I should do? I can try to squash/rebase the git log if that is preferable.

**Comment by parkr** [comment]
This looks just fine to me. Thank you!

@jekyllbot: merge +bug

**Pull request merged**